src/internal/migrations/migrations.go

```diff
@@ -87,12 +88,16 @@ func ApplyMigrations(ctx context.Context, db *pachsql.DB, baseEnv Env, state Sta
 	ctx, end := log.SpanContextL(ctx, "ApplyMigrations", log.InfoLevel)
 	defer end(log.Errorp(&retErr))
 
+	tx, err := db.BeginTxx(ctx, &sql.TxOptions{})
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
 	for _, state := range CollectStates(make([]State, 0, state.n+1), state) {
-		if err := applyMigration(ctx, db, baseEnv, state); err != nil {
+		if err := applyMigration(ctx, tx, baseEnv, state); err != nil {
 			return err
 		}
 	}
-	return nil
+	return errors.EnsureStack(tx.Commit())
```

Comment:
All I would suggest is that we log when Commit fails or the tx rolls back, so the logs are like:

```
applying migration 21: already applied
applying migration 22: ok
applying migration 23: failed due to badness
migrations 22, 23 rolled back, exiting
```